shared/src/test/scala/coop/rchain/shared/MapOpsSpec.scala

```diff
@@ -0,0 +1,32 @@
+package coop.rchain.shared
+
+import org.scalatest._
+import org.scalatest.prop.GeneratorDrivenPropertyChecks
+
+class MapOpsSpec extends FlatSpec with Matchers with GeneratorDrivenPropertyChecks {
+
+  import MapOps._
+
+  "merge" should "retain all keys" in forAll { (m1: Map[String, Int], m2: Map[String, Int]) =>
+    merge(m1, m2, 0, 0).keys should contain theSameElementsAs (m1.keys ++ m2.keys)
+  }
+
+  it should "retain all values" in forAll { (m1: Map[String, Int], m2: Map[String, Int]) =>
+    val result = merge(m1, m2, 0, 0)
+
+    for {
+      key <- result.keys
+    } yield {
+      if (m1.contains(key) && m2.contains(key))
+        result(key) shouldBe ((m1(key), m2(key)))
+      else if (m1.contains(key) && !m2.contains(key))
+        result(key) shouldBe ((m1(key), 0))
+      else if (!m1.contains(key) && m2.contains(key))
+        result(key) shouldBe ((0, m2(key)))
+      else
+        result(key) shouldBe ((0, 0))
+    }
+
+  }
```

Comment:
This property is not readable and it's difficult to get an intuition about the operation by following it.

I am thinking this property can be replaced with something like the following:
1. forall left, right => merge(left,right).size >= left.size
2. forall left, right => merge(left,right).size >= right.size
3. forall left, right => merge(left,right) should contain theSameElementsAs merge(right, left)
4. forall left, right => merge(left,right).map(_.map(_.1)).toMap contains theSameElements as left
...
I'm not sure what the exact properties should be here but one should build the intuition while reading them

Reply:
I find the properties you proposed too weak. They can easily be satisfied with a different implementation which doesn't do what is intended. "not readable" is subject to opinion, I find the assertions clearly stating what the production code is intended to return. Please be more specific

Reply:
The name of the tests doesn't match the implementation. In fact, the chances are you won't find a good name for this test because it tests everything about the zip function.

Regarding the fact that the properties are too weak, the elipsis in my comment was supposed to be replaced with the missing ones :)

This is probably a missing test:
5. forall left, right => merge(left,right).map(.map(.1)).filter(!left.contains(_))